### PR TITLE
[FIX] website_blog, web_editor: refresh options UI after adding a new tag

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] adding a new tag on a blog does not refresh the tag list in
the options tab. That commit sorts the `methodsNames` aplhabetically.
Previously, the unspecified order of the methods just happened to make
the add operation work.

After this commit the tag list is refreshed when a new element is added
to it, in order to not rely on the order of methods anymore.

This PR also fixes three other issues:
- create button now works after hovering suggested tags
- a deleted custom tag can now be re-created
- a race condition during the refreshing of the list of suggested tags

[1]: https://github.com/odoo/odoo/commit/a48a30f954afcb6ff3a59c4f32b05fd0c2cfcd2b

task-2811746

### CODE REVIEW

**Comment by xO-Tx** [comment]
@bso-odoo I still have a weird behaviour when "TAG NAME" input has lost focus [on the last runbot instance]:

https://user-images.githubusercontent.com/25032876/162974303-cfa5fba5-343e-41f2-97ae-baeac924fdf2.mp4

File: addons/website_blog/static/src/js/options.js [hunk 1]
```diff
@@ -121,6 +121,11 @@ options.registry.BlogPostTagSelection = options.Class.extend({
             'display_name': widgetValue,
         };
         this.tagIDs.push(newTagID);
+        return new Promise(resolve => {
+            this.trigger_up('snippet_option_update', {
+                onSuccess: () => resolve(),
+            });
+        });
```
Comment (qsm-odoo) [inline addons/website_blog/static/src/js/options.js:133]: - This has a huge need of comment
- Even if working this is probably way too hacky. Option methods are not supposed to trigger that by themselves nor care about UI updates...
- I don't actually understand it so I cannot suggest alternatives :grimacing: (Need more info, I understand the alphabetical order may be a cause of error but I don't see why in this case, should not the create method be called only anyway?)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -2512,7 +2512,7 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
      */
     getValue(methodName) {
         if (methodName === this.options.createMethod && this.createInput) {
-            return this.createInput._value;
+            return this._isAboutCreate ? this.createInput._value : "";
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2515]: Lacks inline comments too. The commit message seems clear but the code is not straightforward: here I don't understand: "if I am asking for the value  of the create method and there is a create input... return the create input value" is what there was before and seemed very logical. Now, not sure... why can't I get the create input value at any time I am asking for it?

File: addons/website_blog/static/src/js/options.js [hunk 3]
```diff
@@ -107,7 +107,12 @@ options.registry.BlogPostTagSelection = options.Class.extend({
         if (!widgetValue) {
             return;
         }
-        const existing = Object.values(this.allTagsByID).some(tag => tag.name.toLowerCase() === widgetValue.toLowerCase());
+        const existing = Object.values(this.allTagsByID).some(
+            // A tag is already existing only if it was already defined (i.e.
+            // id is a number) or if it appears in the current list of tags.
+            tag => tag.name.toLowerCase() === widgetValue.toLowerCase() &&
+            (typeof(tag.id) === 'number' || this.tagIDs.includes(tag.id))
```
Comment (qsm-odoo) [inline addons/website_blog/static/src/js/options.js:114]: Nitpicking: IMO the arrow on one line notation only is nice if on one line. In this case I would do:
```javascript
stuff.some(tag => {
    return tag_condition A
        && tag_condition B
        [...];
});

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -2560,16 +2560,19 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
     async _search(needle) {
         this._userValueWidgets = this._userValueWidgets.filter(widget => !widget.isDestroyed());
         // Remove select options
-        this._userValueWidgets
-            .filter(widget => {
+        const clearSuggestions = () => {
+            this._userValueWidgets = this._userValueWidgets.filter(widget => !widget.isDestroyed());
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2564]: - Do we still need the first line if you duplicate the line inside here?
- Is this actually necessary in the first place? Don't we have a system that does that automatically on all widgets destroy?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -2560,16 +2560,19 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
...
+            this._userValueWidgets.filter(widget => {
                 return widget instanceof ButtonUserValueWidget &&
-                    widget.el.parentElement.matches('we-selection-items');
+                    widget.el && widget.el.parentElement.matches('we-selection-items');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2567]: Why is this change necessary?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -2590,6 +2593,9 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
         });
 
         await Promise.all(records.slice(0, this.options.limit).map(async record => {
+            // In case of race between updates of the list, make sure to always
+            // start with an empty list of suggestions.
+            clearSuggestions();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2598]: Why can't we just clear the list here only? (once we are ready to update following the RPC?)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 4 above)
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: It is actually needed because of the `button.destroy()` that follows which does not clean up the list. :face_with_spiral_eyes:
I moved the cleanup after.
There is no general "magic" cleanup of that list whenever a destroy happens - but the whole list is flushed on specific case of  `_rerenderXML` after all its content is `destroy`ed.

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 6 above)
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Since the content of the method itself is sequential because of the `await`s, I reordered as follows:
- await rpc
- clear the old list
- await render buttons based on rpc result

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 2 above)
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: @qsm-odoo I replaced this approach by instead restoring the value when the create button is used (to counter the fact that the value is neutralised by the non-create buttons)

File: addons/website_blog/static/src/js/options.js [hunk 7]
```diff
@@ -98,7 +98,10 @@ options.registry.BlogPostTagSelection = options.Class.extend({
      * @see this.selectClass for params
      */
     setTags(previewMode, widgetValue, params) {
-        this.tagIDs = JSON.parse(widgetValue).map(tag => tag.id);
+        if (!this._preventNextSetTagsCall) {
+            this.tagIDs = JSON.parse(widgetValue).map(tag => tag.id);
+        }
+        this._preventNextSetTagsCall = false;
```
Comment (qsm-odoo) [inline addons/website_blog/static/src/js/options.js:104]: I am changing to 
```javascript
        if (this._preventNextSetTagsCall) {
            this._preventNextSetTagsCall = false;
            return;
        }
        this.tagIDs = JSON.parse(widgetValue).map(tag => tag.id);
```
I always prefer that form when possible but in this case it also avoids touching the history of the line we are less likely to refactor later.

File: addons/website_blog/static/tests/tours/blog_tags_tour.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+tour.register('blog_tags', {
+    test: true,
+    url: '/blog',
+}, [{
+        trigger: "article[name=blog_post] a",
+        content: "Go to first blog.",
+    }, {
+        trigger: "section#o_wblog_post_main",
+        content: "Wait for single blog post.",
+        run: () => {}, // it's a check
```
Comment (qsm-odoo) [inline addons/website_blog/static/tests/tours/blog_tags_tour.js:17]: I am changing to:
- Using extra_trigger for this specific second step (I think it's nice to make a dedicated step for steps to "verify / it's a check" but when it's not on a step we actually really want to "verify", I'd say extra_trigger is better to speed things up :+1: 
- Usually I don't think we put a point at the end of "content" for test tours. I'll entertain my OCD and remove them :wink: 
- A while ago, RDE asked me to change to have `content` before `trigger`, I'll do it here too, I agree this is better :)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -2747,6 +2747,10 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
             return;
         }
         if (widget && widget === this.createButton) {
+            // When the create button is clicked, make sure the text
+            // value is restored from the actual input element because
+            // it might have been removed when hovering existing tags.
+            this.createInput._value = this.createInput.el.querySelector('input').value;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2753]: All this createInput/createButton feel very hackish, we'll refactore that someday I hope :open_mouth: This fix is even more hacky although nice catch :+1: I am changing to not virtually emptying the create value on preview updates, I think it's less hackish and seems to work :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -2611,6 +2599,18 @@ const Many2oneUserValueWidget = SelectUserValueWidget.extend({
             method: 'read',
             args: [recTuples.map(([id, _name]) => id), this.options.fields],
         });
+        // Remove select options.
+        this._userValueWidgets.filter(widget => {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2603]: Not able to reproduce the exact issue you described but this is obviously better and I am sure what you describe happens anyway :wink:

**Review by qsm-odoo — COMMENTED** [review]
Also fixing one or 2 types in commit messages but nice ones otherwise :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 9 above)
Comment (bso-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: I tried to do that and did not manage to - I totally missed that `previewMode` was available in the event. :man_facepalming:

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: @bso-odoo Nevermind, I made this: https://github.com/odoo/odoo/commit/edb3b1b6ad6d83263c2258cde74db57fbf327fc3, which works for the current blog usecase but I am not entirely sure it is ok for m2m widgets which could be used with previewMode enabled. I've restored what you did, with a TODO comment, we'll see later :)

**Review by qsm-odoo — APPROVED** [review]
Made [this small diff](https://github.com/odoo/odoo/compare/10498c22b321a15b6ffd2de9af54e0e6948e2149..1bfd9b39c5bb58b912977ddfe41fb1c02fe25fd9) in the end. Thanks for the work :+1: 

Hopefully, we'll get rid of this m2m widget when converting to OWL at some point or even using real m2o widgets :worried: 

@robodoo rebase-ff r+
